File: Lab/C++/BFS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;
int nrNoduri, nrMuchii, nodStart;
vector<int> vizitat;
vector<int> tata;
vector<int> muchii[1001];
void citire(){
    cin >> nrNoduri >> nrMuchii;
    for(int i = 0; i < nrNoduri; ++i){
        vizitat.push_back(0);
        tata.push_back(-1);
    }
    for(int i = 0; i < nrMuchii; ++i){
        int a,b;
        cin >> a >> b;
        muchii[a].push_back(b);
        muchii[b].push_back(a);
    }
    for(int i = 1;i < nrNoduri; ++i)
        sort(muchii[i].begin(),muchii[i].end());
}
void BFS(int start){
    queue <int> coada;
    vizitat[start] = 1;
    coada.push(start);
    while (!coada.empty()){
        int k = coada.front();
        coada.pop();
        for(int i = 0;i < muchii[k].size(); ++i)
            if(vizitat[muchii[k][i]] == 0){
                vizitat[muchii[k][i]] = vizitat[k] + 1;
                tata[muchii[k][i]] = k;
                coada.push(muchii[k][i]);
            }
        cout << k << ' ';
    }
}
int main() {
    citire();
    for(int i = 1; i <= nrNoduri; ++i)
        if( vizitat[i] == 0)
            BFS(i);
    return 0;
}

File: Lab/Python/BFS.py
import queue

f = open("graf.in")
line = f.readline()
v = line.split()

nrNoduri = int(v[0])
nrMuchii = int(v[1])

def citire_matrice_liste(noduri,muchii,orientat):
    l = {}
    for i in range(1, noduri + 1):
        l[i] = []

    for i in range(muchii):
        line = f.readline()
        v = line.split()
        if orientat:
            l[int(v[0])].append(int(v[1]))
        else:
            l[int(v[0])].append(int(v[1]))
            l[int(v[1])].append(int(v[0]))
    return l

lista = citire_matrice_liste(nrNoduri,nrMuchii,0)
f.close()

vizitat = [0] * (nrNoduri + 1)
tata = [0] * (nrNoduri + 1)
dist = [None] * (nrNoduri + 1)
parcurg = []
def BFS(start):
    global lista, tata, vizitat, dist, parcurg
    q = []
    q.append(start)
    vizitat[start] = 1
    dist[start] = 0
    while len(q) > 0:
        x = q.pop(0)
        parcurg.append(x);
        for j in lista[x]:
            if vizitat[j] == 0:
                q.append(j)
                vizitat[j] = 1
                tata[j] = x
                dist[j] = dist[x] + 1

def afisare_listant(start):
    while start!=0:
        print(start, end = " ")
        start = tata[start]

start = 1
BFS(start)
print(vizitat)
print(tata)
print(dist)
print(parcurg)
